spec.go

```diff
@@ -179,71 +177,95 @@ type ImageConfig struct {
-	// Args are the build args to pass to the build.
-	Args map[string]string `yaml:"args,omitempty" json:"args,omitempty"`
-	// File is the path to the build file in the build context
+// SourceBuild is used to generate source from a DockerFile build, either
+// inline or from a local file.
+type SourceBuild struct {
+	// A source specification to use as the context for the Dockerfile build
+	Source Source `yaml:"source,omitempty" json:"source,omitempty"`
+
+	// DockerFile is the path to the build file in the build context
 	// If not set the default is assumed by buildkit to be `Dockerfile` at the root of the context.
 	// This is exclusive with [Inline]
-	File string `yaml:"file,omitempty" json:"file,omitempty"`
-
+	DockerFile string `yaml:"dockerfile,omitempty" json:"dockerfile,omitempty"`
 	// Inline is an inline build spec to use.
 	// This can be used to specify a dockerfile instead of using one in the build context
 	// This is exclusive with [File]
 	Inline string `yaml:"inline,omitempty" json:"inline,omitempty" jsonschema:"example=FROM busybox\nRUN echo hello world"`
+
+	// Target specifies the build target to use.
+	// If unset, the default target is determined by the frontend implementation
+	// (e.g. the dockerfile frontend uses the last build stage as the default).
+	Target string `yaml:"target,omitempty" json:"target,omitempty"`
+	// Args are the build args to pass to the build.
+	Args map[string]string `yaml:"args,omitempty" json:"args,omitempty"`
+}
+
+// Command is used to execute a command to generate a source from a docker image.
+type Command struct {
+	// Dir is the working directory to run the command in.
+	Dir string `yaml:"dir,omitempty" json:"dir,omitempty"`
+
+	// Mounts is the list of sources to mount into the build steps.
+	Mounts []SourceMount `yaml:"mounts,omitempty" json:"mounts,omitempty"`
+
+	// List of CacheDirs which will be used across all Steps
+	CacheDirs map[string]CacheDirConfig `yaml:"cache_dirs,omitempty" json:"cache_dirs,omitempty"`
+
+	// Env is the list of environment variables to set for all commands in this step group.
+	Env map[string]string `yaml:"env,omitempty" json:"env,omitempty"`
+
+	// Steps is the list of commands to run to generate the source.
+	// Steps are run sequentially and results of each step should be cached.
+	Steps []*BuildStep `yaml:"steps" json:"steps" jsonschema:"required"`
+}
+
+// Source defines a source to be used in the build.
+// A source can be a local directory, a git repositoryt, http(s) URL, etc.
+type Source struct {
+	// This is an embedded union representing all of the possible source types.
+	// Exactly one must be non-nil, with all other cases being errors.
+	//
+	// === Begin Source Variants ===
+	DockerImage *SourceDockerImage `yaml:"image,omitempty" json:"image,omitempty"`
+	Git         *SourceGit         `yaml:"git,omitempty" json:"git,omitempty"`
+	HTTPS       *SourceHTTPS       `yaml:"https,omitempty" json:"https,omitempty"`
```

Comment:
Nit: this should probably just be `http`.
Tried to think of a better name than that but am coming up short.

Reply:
The field name? The Struct name? The json tag? All of the above?

Reply:
All of the above.

Reply:
done